### PR TITLE
NET-151 Seeker Ownership

### PR DESCRIPTION
**Pull request by JCSanPedro**
## Summary

This PR integrates the Cennznet State Oracle Bridge, and addresses the problem of verifying seeker ownership for when Nodes join epochs.

## Motivation

Owning a seeker is required to run a node. This is complicated because of the following reasons:
- Seekers exist on the ethereum mainnet and the sylo event protocol will run on Cennznet EVM
- The ethereum account the Node operates is solely owned by the Node, and is expected to be different to the user's mainnet wallet (which owns the seeker). It is also expected that the mainnet wallet does not need to perform any cennznet evm transaction, as this would require bridging Sylo/CPAY to both Node account and the wallet account.

### State Oracle Integration

An additional contract; [Seekers.sol](https://github.com/dn3010/sylo-ethereum-contracts/compare/NET-139-additional-events...NET-151-seeker-ownership?expand=1#diff-980e168178b8a49979d8ed1d97a206ec371a1a7e2c44fd757cd6dc0ff4f5d055) is implemented to manage reads over the state bridge.

The function [requestVerifitcation()](https://github.com/dn3010/sylo-ethereum-contracts/compare/NET-139-additional-events...NET-151-seeker-ownership?expand=1#diff-980e168178b8a49979d8ed1d97a206ec371a1a7e2c44fd757cd6dc0ff4f5d055R81) is available to be called to request the oracle to check which address owns a specific seeker.

Once the oracle has performed that check, it will invoke the callback [confirmOwnership()](https://github.com/dn3010/sylo-ethereum-contracts/compare/NET-139-additional-events...NET-151-seeker-ownership?expand=1#diff-980e168178b8a49979d8ed1d97a206ec371a1a7e2c44fd757cd6dc0ff4f5d055R104), which will then store the owner of that seekerId within a mapping in `Seekers.sol`.

That mapping can then be queried with [ownerOf(uint256 seekerId)](https://github.com/dn3010/sylo-ethereum-contracts/compare/NET-139-additional-events...NET-151-seeker-ownership?expand=1#diff-980e168178b8a49979d8ed1d97a206ec371a1a7e2c44fd757cd6dc0ff4f5d055R116) to retrieve the confirmed owner of the seekerId.

Once confirmation has occurred, the node can call [setSeekerId()](https://github.com/dn3010/sylo-ethereum-contracts/compare/NET-139-additional-events...NET-151-seeker-ownership?expand=1#diff-8e6a3b08df2f10630c0ad1aa5b9b0e8b495f48f7faa131b5558af9d77d8012bdR109) to create an association between the Node and a seeker Id (and the account which owns the Seeker).

Confirmations of ownership are only valid for a certain number of blocks. Once a confirmation has expired, any calls to `ownerOf()` will return `address(0)`. By limiting a valid duration, we can ensure that Node operators actually own the seeker at the time the Node joins the epoch or initially sets the seeker Id.

Node operators are expected to call `requestVerification()` prior to making a transaction that requires the Node to own a seeker. This currently includes `joinEpoch()` and `setSeekerId()`. Failing to do will cause the previously mentioned transactions to revert (since the confirmation of ownership will have expired).

### Node/Seeker account verification

When a Node sets the seekerId/account via `setSeekerId()`, the contract needs to verify that the user operating the Node also owns the wallet which owns the seekers.

This is achieved by having the seeker wallet sign a proof which is then also additionally submitted as a parameter to `setSeekerId()`. The proof includes the seeker Id, the node's address, a human readable prefix string, and also a block number which represents when the proof was generated. The block number prevents re-use of the proof as the proof will only be valid within a 100 blocks of when the proof was generated.

The contract will verify the signature of the proof to confirm the node operator has access to the seeker wallet.

## Sequence Diagram

A [sequence diagram is available](https://github.com/dn3010/sylo-ethereum-contracts/blob/NET-151-seeker-ownership/docs/StateOracleSequence.png), which maps out all interactions between the user/wallet, contracts and networks for the state oracle integration.

## Checklist:

If not applicable to this PR, the item can be checked, crossed out, or removed.

### Documentation

- [x] Updated the readme documents (`README.md`, etc.)
- [x] Updated the spec if applicable (`docs/spec.md`, etc.)
- [x] Updated any doc comments, and checked for spelling mistakes.

### Local Build

- [x] Removed references to any local packages in build files
- [x] Ran `npm install` and no changes were made to `package-lock.json`
- [x] Ran `npm test` with no failures
- [ ] Ran `npm coverage` and validated coverage has not decreased

### GitHub

- [ ] Base branch has been set correctly
- [x] PR has been assigned an owner
- [x] There are no conflicts with the base branch
- [x] Any dependent changes have been merged and published in upstream modules
- [x] Performed a self-review of my code
- [x] All commits are signed
- [x] Additional reviews have been performed
- [x] Removed the WIP message from the top of this PR

### CircleCI

- [x] Tests are passing
- [ ] Test coverage report has been reviewed

### CODE REVIEW

File: contracts/Seekers.sol [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        require(success, "oracle request failed");
+
+        uint256 requestId = abi.decode(returnData, (uint256));
+
+        requests[requestId] = seekerId;
+    }
+
+
+    function confirmOwnership(uint256 requestId, uint256 timestamp, bytes32 returnData) external {
+        require(msg.sender == _oracle, "must be state oracle");
```
Comment (teinnt) [inline contracts/Seekers.sol:142]: I think we can say

```suggestion
        require(msg.sender == _oracle, "msg.sender must be state oracle");
```

File: contracts/Seekers.sol [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        if (owner.expiry < block.number) {
+            return address(0);
+        }
+
+        return owner.owner;
+    }
+
+    function withdrawAll(address recipient) public onlyOwner {
+        uint256 balance = address(this).balance;
+        payable(recipient).transfer(balance);
```
Comment (teinnt) [inline contracts/Seekers.sol:170]: The msg.sender (or owner) will pay fee for this transaction right?

File: contracts/Seekers.sol [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    function setCallbackGasLimit (uint256 _callbackGasLimit) external onlyOwner {
+        callbackGasLimit = _callbackGasLimit;
+    }
+
+    function setCallbackBounty (uint256 _callbackBounty) external onlyOwner {
+        callbackBounty = _callbackBounty;
+    }
+
+    function requestVerification(uint256 seekerId) external payable {
```
Comment (teinnt) [inline contracts/Seekers.sol:94]: Does this `payable` here mean the `Seekers` contract can pay for the `Oracle` contract? Or at the moment the sender will pay for it?

File: deploy/nikau_deployment_phase_two.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+ "token": "0xcccCccCC000042B4000000000000000000000000",
+ "listings": "0xfA5227d68c141be0855d7447fFA350BDf7d13B12",
+ "ticketing": "0x7BA00106063d256EA3DC5C7223aF1DE86aDe042e",
+ "ticketingParameters": "0x39f826f4b33568D68159058305D10C15307D00f9",
+ "directory": "0x773F81E0bCFC64a545c2F61B7556BE13ABF5a90a",
+ "rewardsManager": "0xf4374ee1B2aE4c4667568f92a9236B254e5EA67f",
+ "epochsManager": "0x841D64d2c9AF99841dA6410329a3805BD95d2A0F",
+ "stakingManager": "0x92CF9a0D88922e5F0943325A052a26f01Fa4c484",
+ "seekers": "0xce0eB06C84Ed237a5C3270262afd114c7E94B23F"
+}
```
Comment (teinnt) [inline deploy/nikau_deployment_phase_two.json:12]: I guess we should update the file writer to add the last line after saving contract addresses.

**Review by teinnt — COMMENTED** [review]
Thank you for the diagram and comments throughout the code, they are very useful for me to understand. You may see a lot of comments, that is because I am over checking, none of them relate to the logic. Your logic is very strong. I really (super) like the the seeker owner's temporarily check. I've just finished 1/3 of the PR and will continue on Monday!

File: test/listings.ts [hunk 5]
```diff
@@ -57,9 +66,129 @@ describe('Listing', () => {
...
+      accounts[1],
+      1,
+    );
+
+    await listings.connect(seekerAccount).revokeSeekerAccount(owner);
+
+    const listing = await listings.getListing(owner);
+
+    expect(listing.seekerAccount).to.equal(
+      '0x0000000000000000000000000000000000000000',
```
Comment (teinnt) [inline test/listings.ts:170]: I saw that we can also use `ethers.constants.AddressZero`, but it really depends on your opinion and preference.

```suggestion
      ethers.constants.AddressZero,
```

**Review by teinnt — COMMENTED** [review]
That's all I can find. Otherwise your tests look very awesome!

File: contracts/Seekers.sol [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        uint256 balance = address(this).balance;
+        payable(recipient).transfer(balance);
+    }
+
+    function withdrawAllFee(address recipient) public onlyOwner {
+        IERC20 token = IERC20(_token);
+        uint256 amount = token.balanceOf(address(this));
+        token.transfer(recipient, amount);
+    }
+}
```
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply contracts/Seekers.sol:178]: Yeah we can probably add a .sol prettier in another PR to automatically catch these things

File: contracts/Listings.sol [hunk 7]
```diff
@@ -18,20 +26,30 @@ contract Listings is Initializable, OwnableUpgradeable {
         //  { name: string, multiaddrs: string[] }
         string publicEndpoint;
 
+        // The account which owns a seeker that will be used to
+        // operate the Node for this listing.
+        address seekerAccount;
+
+        // The id of the seeker used to operate the node. The owner
+        // of this id should be the seeker account.
+        uint256 seekerId;
```
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply contracts/Listings.sol:35]: No, want to keep `uint256` as erc721 tokenIds are `uint256`.

File: contracts/Seekers.sol [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    mapping (uint256 => uint256) public requests;
+    mapping (uint256 => Owner) private owners;
+
+    /**
+     * @notice The address of the Seekers NFT contract on ethereum mainnet.
+     */
+    address public _seekers;
+
+    /**
+     * @notice The address of the token used for fees.
```
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply contracts/Seekers.sol:25]: It's actually for paying the oracle but I'll update the comment appropriately.

File: contracts/Listings.sol [hunk 9]
```diff
@@ -46,8 +64,12 @@ contract Listings is Initializable, OwnableUpgradeable {
      */
     uint16 public defaultPayoutPercentage;
 
-    function initialize(uint16 _defaultPayoutPercentage) external initializer {
+    function initialize(
+        Seekers seekers,
+        uint16 _defaultPayoutPercentage
```
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply contracts/Listings.sol:69]: Good point. We can review the style guide and update all of our styling to match in a seperate PR.

File: contracts/Seekers.sol (hunk 2 above)
Comment (JCSanPedro) [reply contracts/Seekers.sol:170]: Yes 👍

File: contracts/Seekers.sol (hunk 3 above)
Comment (JCSanPedro) [reply contracts/Seekers.sol:94]: `payable` allows the sender to transfer CPAY to the contract in the same transaction, so that the contract can pay for it.

File: contracts/Seekers.sol [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        (bool success, bytes memory returnData) = _oracle.call(remoteCallRequest);
+        require(success, "oracle request failed");
+
+        uint256 requestId = abi.decode(returnData, (uint256));
+
+        requests[requestId] = seekerId;
+    }
+
+
+    function confirmOwnership(uint256 requestId, uint256 timestamp, bytes32 returnData) external {
```
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply contracts/Seekers.sol:141]: The `timestamp` is used by the Oracle when calling this function, so we need to keep it as part of the function args, we just don't use the timestamp ourselves.

File: deploy/genesis.config.ts [hunk 11]
```diff
@@ -33,10 +41,27 @@ type ContractParameters = {
...
+    seekersERC721:
+      hre.network.name === 'rata'
+        ? '0x856D2df6998AcA9FDC4B2CA316b0527081cee8DD'
+        : hre.network.name === 'nikau'
+        ? '0x49e5706c06a263ac142d0a8c06f7ab247c1d978a'
+        : '',
+    oracle: '0x0000000000000000000000000000000000006bb4',
+    validDuration: 200,
+    callbackGasLimit: 300000,
+    callbackBounty: hre.ethers.utils.parseEther('2'),
```
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply deploy/genesis.config.ts:62]: In their EVM it will be correctly converted to 2 CPAY.

File: test/listings.ts (hunk 5 above)
Comment (JCSanPedro) [reply test/listings.ts:170]: This is better I think

**Review by JoshD641 — APPROVED** [review]
It looks good to me - I looked for edge cases around payment and changing seekers, but couldn't find any issues.

**Review by JoshD641 — APPROVED** [review]

**Review by teinnt — APPROVED** [review]
LGTM!
